### ISSUE TITLE
Investigate adopting react toolkit

### ISSUE THREAD

**Issue by nrfox** [kiali/kiali#7110]
### What do you want to improve?

Investigate what it would take to use react-toolkit on the frontend and provide some analysis on pros/cons and level of effort to integrate. Any new integration must work with OSSMC.

**Comment by nrfox** [comment]
Some investigation was done awhile back and that PR can be used as a reference: https://github.com/kiali/kiali-ui/pull/2267.

@ferhoyos are there any potential risks wrt OSSMC?

**Comment by ferhoyos** [comment]
I took a look to redux toolkit in the past and it is great to remove redux boilerplate and simplify the code. Wrt to OSSMC we need to check that redux and redux-toolkit can coexist in the same application and access to the same state (Openshift uses redux as state management library). If possible, I suggest to do a small PoC in Kiali and test it in OSSMC to check compatibility.

**Comment by jshaughn** [comment]
Does this tie into our current Epic work?

**Comment by ferhoyos** [comment]
No, I don't think so. Redux-toolkit does not necessarily increase the performance of the application. It is more focused to simplify redux code and automate some error-prone manual stuff.

These are the main reasons to migrate from redux to redux-toolkit (https://redux.js.org/introduction/why-rtk-is-redux-today):
- Redux Toolkit simplifies store setup down to a single clear function call, while retaining the ability to fully configure the store's options if you need to
- Redux Toolkit eliminates accidental mutations, which have always been the main cause of Redux bugs
- Redux Toolkit eliminates the need to write any action creators or action types by hand
- Redux Toolkit eliminates the need to write manual and error-prone immutable update logic
- Redux Toolkit makes it easy to write a Redux feature's code in one file, instead of spreading it across multiple separate files
- Redux Toolkit offers excellent TS support, with APIs that are designed to give you excellent type safety and minimize the number of types you have to define in your code
- RTK Query can eliminate the need to write any thunks, reducers, action creators, or effect hooks to manage fetching data and tracking loading state

**Comment by nrfox** [comment]
> No, I don't think so. Redux-toolkit does not necessarily increase the performance of the application.

Beyond just making it easier to develop on the frontend, I'd actually argue that it is related to perf. Some pages will make multiple calls to the backend for the same information simply because they are not sharing state. Especially navigating between pages there's not much state shared today.

You can see in the PR I linked that the investigation for standalone Kiali is basically done. It'd just be a matter of ensuring rtk works with OSSMC. The nice thing about integrating this is that it can be done incrementally over time instead of all at once. It's the same redux the frontend is using today with some additional tooling on top.

**Comment by jshaughn** [comment]
For me it's all about bang for the buck.  Refactors take time and introduce churn into working code.  But, it can be worth it as long as the reward is worth the effort.  Just convince me.

**Comment by nrfox** [comment]
> Here we have two different things:

Agreed these are separate things and adopting rtk won't increase performance but it should make it easier to use/re-use redux across the app which will make performance better by requiring less calls to the backend to refetch the same info.

> IMO I would follow this approach (1 and 2 can be done in parallel):

That sounds good to me!

> For me it's all about bang for the buck. Refactors take time and introduce churn into working code. But, it can be worth it as long as the reward is worth the effort. Just convince me.

IMO this is a pretty low cost refactor as it can happen gradually over time and based on the reference PR above it will mostly involve removing boilerplate. It falls in line with making the app feel better to use and less clunky since page transitions could require less loading screens if we can share more state between them.
